## Results

### A strategy for incorporating heterogeneous clinical measurements

Text.

![**Structure data decomposition can integrate clinical measurements with varying degrees of overlap.** A) General description of the data. Subjects with a MRSA infection had samples collected at admission, and then were monitored during their disease course. Identified subjects in the study were XXXX matched. Samples from each subject included serum, plasma, and isolated PBMCs. Inidividual subjects had cytokine measurements from either serum, plasma, or both. Similarly, PBMC gene expression was available for many but not all subjects. These measurements were then used to predict disease outcome, defined as the infection resolving (ARMB) or persisting (APMB) within XXX months. B) Overall structure of the data. Cytokine measurements from either plasma or serum can be arranged in a three-dimensional tensor, wherein one dimension indicates subject, cytokine, or sample source, each. In parallel, gene expression measurements are aligned with cytokine measurements along the subject dimension but not across genes. C) Data reduction is performed by identifying additively-separable components represented by the outer product of vectors along each dimension. The subjects dimension is shared across both the tensor and matrix reconstruction. Venn diagram of the variance explained by each factorization method. Canonical polyadic (CP) decomposition can explain the variation present within either the antigen-specific tensor or glycan matrix on their own [@PMID:18003902]. CMTF allows one to explain the shared variation between the matrix and tensor [@PMID:31251750]. In contrast, here we wish to explain the total variation across both the tensor and matrix. This is accomplished with TMTF [@DOI:10.1101/2021.01.03.425138].](figure1.svg "Figure 1"){#fig:figure1 width="100%"}

### TMTF factorization recovers consistent patterns across data types

Text.

![**TMTF factorization recovers consistent patterns across data types.** A) Number of components used in the TMTF decomposition versus the percent variance reconstructed (R2X). Bâ€“D) Decomposed components along the subjects (B), cytokines (C), and sample source (D) dimensions. Subjects are labeled according to the outcome of their infection and the cohort from which they were derived.](figure2.svg "Figure 2"){#fig:figure2 width="100%"}

### Factorization-based data fusion improves predictions of patient outcomes

Text.

![**Factorization-based data fusion improves predictions of patient outcomes.** AAA. BBB. CCC.](figure3.svg "Figure 3"){#fig:figure3 width="100%"}

### Section 4

Text.

![**Figure 4.** AAA. BBB. CCC.](figure4.svg "Figure 4"){#fig:figure4 width="100%"}

### Section 5

Text.

![**Figure 5.** AAA. BBB. CCC.](figure5.svg "Figure 5"){#fig:figure5 width="100%"}

### Section 6

Text.

![**Figure 6.** AAA. BBB. CCC.](figure6.svg "Figure 6"){#fig:figure6 width="100%"}
