## Results

### A tensor-based strategy for integrating heterogeneous clinical measurements

To identify common patterns across our cytokine and RNAseq measurements, we first sought to identify a consistent pattern in the measurements. We determined that all our measurements were of three types—plasma cytokines, serum cytokines, and RNAseq from isolated PBMCs ({@fig:figure1}A). Known differences exist in cytokine measurements between plasma and serum, however we can expect to see certain shared variation across subjects. Therefore, while not every subject had every type of measurement, the study contained two types of measurements that can be aligned across cytokines and subjects, and then RNAseq measurements that only share a subject dimension.

In situations where measurements can be aligned along two or more dimensions, generalizations of matrix decomposition methods exist for data in tensor form (CITE). We started by structuring our data into a 3-mode tensor, with subject, cytokine, and cytokine source axes ({@fig:figure1}B). This was paired with the RNAseq measurements in a matrix with axes representing genes and subjects. To integrate each data type, we used coupled matrix-tensor factorization (CMTF). CMTF concatenates unfolded datasets during alternating least squares solving of a coupled dimension (in our case the subjects). In doing so, it solves for an optimal low-rank approximation of both datasets ({@fig:figure1}C). CMTF maximizes the variance explained across both datasets; other approaches can maximize explanation of other subsets of variance, including canonical polyadic (CP) decomposition for just the cytokines tensor, principal components analysis (PCA) for just the RNAseq matrix, or partial least squares regression in tensor form (tPLS) for specifically the shared variance ({@fig:figure1}D).

![**Structure data decomposition can integrate clinical measurements with varying degrees of overlap.** A) General description of the data. Subjects with a MRSA infection had samples collected at admission, and then were monitored during their disease course. Identified subjects in the study were XXXX matched. Samples from each subject included serum, plasma, and isolated PBMCs. Inidividual subjects had cytokine measurements from either serum, plasma, or both. Similarly, PBMC gene expression was available for many but not all subjects. These measurements were then used to predict disease outcome, defined as the infection resolving (ARMB) or persisting (APMB) within XXX months. B) Overall structure of the data. Cytokine measurements from either plasma or serum can be arranged in a three-dimensional tensor, wherein one dimension indicates subject, cytokine, or sample source, each. In parallel, gene expression measurements are aligned with cytokine measurements along the subject dimension but not across genes. C) Data reduction is performed by identifying additively-separable components represented by the outer product of vectors along each dimension. The subjects dimension is shared across both the tensor and matrix reconstruction. Venn diagram of the variance explained by each factorization method. Canonical polyadic (CP) decomposition can explain the variation present within either the antigen-specific tensor or glycan matrix on their own [@PMID:18003902]. CMTF allows one to explain the shared variation between the matrix and tensor [@PMID:31251750]. In contrast, here we wish to explain the total variation across both the tensor and matrix. This is accomplished with CMTF [@DOI:10.1101/2021.01.03.425138].](figure1.svg "Figure 1"){#fig:figure1 width="100%"}

### CMTF parameter tuning to predict infection outcome

CMTF requires two input parameters that influence the resulting decomposition. First, decomposition can be performed using a varying number of components. Second, CMTF aims to explain the maximum variance possible. Because our cytokine and RNAseq measurements do not have comparable units, we allowed the scaling between both datasets to vary. As expected, CMTF led to a striking reduction in the data while explaining a large fraction of its variance—10 components were sufficient to explain roughly 80% of the variance ({@fig:params}A). (COMPARE TO PCA?) The resulting factors were responsive to the relative scale of each data type ({@fig:params}B). As expected, the data with larger scaling was more readily explained by the dimensionality reduction.

We chose to prioritize CMTF results that were able to predict infection outcome most accurately. To do so, we used the subject factors to predict outcome using logistic regression. Prediction accuracy was quantified using 10-fold cross-validation. Briefly, 10% of the subjects were left out from the logistic regression model, and then the remaining 90% were used to learn the relationship between each factor and outcome. Repeating this process across all groups of subjects provided predictions for every subject. Using this process with varying settings within CMTF, we observed a peak prediction performance at 9 components ({@fig:params}C) and a relative cytokine scale of 32 ({@fig:params}D).

![**CMTF parameter tuning to predict infection outcome.** A) Number of components used in the CMTF decomposition versus the percent variance reconstructed (R2X). B) Accuracy upon reconstruction (R2X) of entire data and RNA or cytokine measurements separately with varying scaling between the two datasets. C–D) Balanced prediction accuracy in predicting infection outcome with varying number of components (C) and scaling (D).](figure2.svg "Figure 2"){#fig:params width="100%"}

### CMTF-based data integration improves predictions of patient outcomes

Text.

![**CMTF-based data integration improves predictions of patient outcomes.** A) Prediction accuracy when using the cytokines or gene expression measurements on their own, or when combined using CMTF. Accuracy was assessed by cross-validation within the training cohort. B) Prediction ROC curve based on cross-validation within the training cohort. C) ROC curve from validation cohort. C) Prediction accuracy using all pairs of components from CMTF. D) The prediction decision function with respect optimal pair of components for prediction.](figure3.svg "Figure 3"){#fig:figure3 width="100%"}

### MRSA infection response involves combined immune cell-cytokine feedback

Text.

![**A combined immune cell-cytokine regulatory program.** AAA. BBB. CCC.](figure4.svg "Figure 4"){#fig:figure4 width="100%"}

### Figure 5

Text.

![**A combined immune cell-cytokine regulatory program.** AAA. BBB. CCC.](figure5.svg "Figure 5"){#fig:figure5 width="100%"}

### Figure 6

Text.

![**A combined immune cell-cytokine regulatory program.** AAA. BBB. CCC.](figure6.svg "Figure 6"){#fig:figure6 width="100%"}