## Methods

### Patients and sample collection

This case-controlled study consisted of 58 SAB patients (28 APMB and 30 ARMB) propensity matched by sex, race, age, haemodialysis status, Type I diabetes, presence of an implantable device.  Details of clinical characteristics of study cohort are presented in Table 1. SAB cases were evaluated and consented for enrolment in the S. aureus Bacteremia Group (SABG) biorepository at Duke University Medical Centre (DUMC). Cases for the current study were carefully selected based on the following inclusion criteria: laboratory confirmed MRSA bacteremia; received appropriate vancomycin therapy; enrolled in the SABG study between 2007 and 2017 (to ensure contemporary medical practices). APMB was defined as patients had continuous MRSA positive blood cultures for at least 5 days after vancomycin antibiotic treatment [@DOI:10.1073/pnas.1909849116]; while ARMB patients had initial blood cultures that were positive for MRSA, but the subsequent blood cultures were negative.

### Molecular analysis

#### Luminex-based cytokine measurement

Human 38-plex magnetic cytokine/chemokine kits (EMD Millipore, HCYTMAG-60K- PX38) were used per manufacturer’s instructions. The panel includes IL-1RA, IL-10, IL-1α, IL- 1β, IL-6, IFN-α2, TNF/TNF-α, TNF-β/LT-α, sCD40L, IL-12p40, IFN-γ, IL-12/IL-12p70, IL-4, IL-5, IL-13, IL-9, IL-17A, GRO/CXCL1, IL-8/CXCL8, eotaxin/CCL11, MDC/CCL22, fractalkine/CX3CL1, IP-10/CXCL10, MCP-1/CCL2, MCP-3/CCL7, MIP-1α/CCL3, MIP-1β/CCL4, IL-2, IL-7, IL-15, GM-CSF, Flt-3L/CD135, G-CSF, IL-3, EGF, FGF-2, TGF-α, and VEGF. Fluorescence was quantified using a Luminex 200TM instrument. Cytokine/chemokine concentrations were calculated using Milliplex Analyst software version 4.2 (EMD Millipore). Luminex assay and analysis were performed by the UCLA Immune Assessment Core.

#### RNA sequencing, mapping, quantifications and quality control

Total RNA was isolated with Qiagen RNA Blood kit, and quality control was performed with Nanodrop 8000 and Agilent Bioanalyzer 2100. Globin RNA was removed with Life Technologies GLOBINCLEAR (human) kit. Libraries for RNA-Seq were prepared with KAPA Stranded mRNA-Seq Kit. The workflow consists of mRNA enrichment, cDNA generation, and end repair to generate blunt ends, A-tailing, adaptor ligation and PCR amplification. Different adaptors were used for multiplexing samples in one lane. Sequencing was performed on Illumina Hiseq3000 for a single read 50 run. Each sample generated an average of 15 million reads. Data quality check was done on Illumina SAV. Demultiplexing was performed with Illumina Bcl2fastq2 v 2.17 program.

### Computational Analysis

#### Data Factorization

The data tensor was constructed by creating two matrix slices consisting of clinical cytokine measurements and RNA sequencing counts for all patients. The two slices were separately normalized prior to factorization. <!-- TODO: Insert short description of normalization here. --> PARAFAC2 decomposition was performed using the TensorLy Python package [TODO: CITE].

<!-- TODO: Add more about the math details of PARAFAC2. -->

#### Gene expression deconvolution

Following the protocol of the R package "immunedeconv" [@PMID:31510660; @PMID:32124323], we used a matrix of TPM values as input to perform the deconvolution. The read counts per gene were normalized into transcripts per kilobase million (TPM) reads mapped for the human genes. For this analysis, we considered only genes with >10 counts per million in 10 or more samples as expressed and tested a total of 12,253 human genes (out of 58,870 human genes). Deconvolution was then performed using CIBERSORT using the LM22 signatures matrix file consisting of 547 genes that accurately distinguish 22 mature human hematopoietic populations isolated from peripheral blood or in vitro culture conditions, including seven T cell types, naive and memory B cells, plasma cells, NK cells, and myeloid subsets [@PMID:31510660].  
